### PR TITLE
[SAC-42] Unify Spark ML and Dataframe operator into a uniform Spark Process

### PR DESCRIPTION
**Pull request by merlintang**
This PR is used to combine Spark ML and DF operation into the same Spark Process. 
For example: 
Give a Spark application with ML and DF job, they are recorded into a same Spark Process in Atlas. 
screen shot 2018-04-23 at 6 22 48 pm

### CODE REVIEW

File: spark-atlas-connector/src/main/scala/com/hortonworks/spark/atlas/ml/MLPipelineEventProcessor.scala [hunk 1]
```diff
@@ -18,27 +18,25 @@
 package com.hortonworks.spark.atlas.ml
 
 import scala.collection.mutable
-
```
Comment (jerryshao) [inline spark-atlas-connector/src/main/scala/com/hortonworks/spark/atlas/ml/MLPipelineEventProcessor.scala]: Revert this change, should have a blank line.

File: spark-atlas-connector/src/main/scala/com/hortonworks/spark/atlas/ml/MLPipelineEventProcessor.scala [hunk 2]
```diff
@@ -18,27 +18,25 @@
...
 import org.apache.atlas.model.instance.AtlasEntity
 import org.apache.spark.ml._
 import org.apache.spark.scheduler.SparkListenerEvent
 import org.apache.spark.sql.Dataset
 import org.apache.spark.sql.execution.FileRelation
 import org.apache.spark.sql.execution.datasources.LogicalRelation
 import org.apache.spark.sql.types.{IntegerType, StringType, StructType}
-
```
Comment (jerryshao) [inline spark-atlas-connector/src/main/scala/com/hortonworks/spark/atlas/ml/MLPipelineEventProcessor.scala]: ditto.

File: spark-atlas-connector/src/main/scala/com/hortonworks/spark/atlas/ml/MLPipelineEventProcessor.scala [hunk 3]
```diff
@@ -18,27 +18,25 @@
...
 import com.hortonworks.spark.atlas._
 import com.hortonworks.spark.atlas.types.{AtlasEntityUtils, external, internal}
 import com.hortonworks.spark.atlas.utils.Logging
 import com.hortonworks.spark.atlas.utils.CatalogUtils
+import org.apache.spark.sql.catalyst.catalog.HiveTableRelation
+import org.apache.spark.sql.catalyst.plans.logical.View
```
Comment (jerryshao) [inline spark-atlas-connector/src/main/scala/com/hortonworks/spark/atlas/ml/MLPipelineEventProcessor.scala]: Please fix the import ordering, we followed Spark's scalastyle.

File: spark-atlas-connector/src/main/scala/com/hortonworks/spark/atlas/sql/CommandsHarvester.scala [hunk 4]
```diff
@@ -69,18 +67,36 @@ object CommandsHarvester extends AtlasEntityUtils with Logging {
...
 
-      // create process entity
+
       val inputTablesEntities = inputsEntities.flatMap(_.headOption).toList
       val outputTableEntities = List(outputEntities.head)
-      val pEntity = processToEntity(qd.qe, qd.executionId, qd.executionTime, inputTablesEntities,
-        outputTableEntities, qd.query)
-      Seq(pEntity) ++ inputsEntities.flatten ++ outputEntities
+
+      val logMap = Map(
```
Comment (jerryshao) [inline spark-atlas-connector/src/main/scala/com/hortonworks/spark/atlas/sql/CommandsHarvester.scala]: Extract the common logic to a method for here and below.

File: spark-atlas-connector/src/main/scala/com/hortonworks/spark/atlas/sql/CommandsHarvester.scala [hunk 5]
```diff
@@ -69,18 +67,36 @@ object CommandsHarvester extends AtlasEntityUtils with Logging {
...
+        Seq(pEntity) ++ inputsEntities.flatten ++ outputEntities
+      }
+
     }
   }
 
   object InsertIntoHadoopFsRelationHarvester extends Harvester[InsertIntoHadoopFsRelationCommand] {
     override def harvest(node: InsertIntoHadoopFsRelationCommand, qd: QueryDetail)
-        : Seq[AtlasEntity] = {
+    : Seq[AtlasEntity] = {
```
Comment (jerryshao) [inline spark-atlas-connector/src/main/scala/com/hortonworks/spark/atlas/sql/CommandsHarvester.scala]: revert this style change.

File: spark-atlas-connector/src/test/scala/com/hortonworks/spark/atlas/sql/CreateHiveTableAsSelectHarvesterSuite.scala [hunk 6]
```diff
@@ -198,7 +198,7 @@ class CreateHiveTableAsSelectHarvesterSuite extends FunSuite with Matchers with
     val pEntity = entities.head
 
     pEntity.getAttribute(AtlasClient.REFERENCEABLE_ATTRIBUTE_NAME) should be (
-      sparkSession.sparkContext.applicationId + ".0")
+     sparkSession.sparkContext.applicationId)
```
Comment (jerryshao) [inline spark-atlas-connector/src/test/scala/com/hortonworks/spark/atlas/sql/CreateHiveTableAsSelectHarvesterSuite.scala]: two spaces.

File: spark-atlas-connector/src/main/scala/com/hortonworks/spark/atlas/types/internal.scala [hunk 7]
```diff
@@ -217,4 +219,50 @@ object internal extends Logging {
     entity.setAttribute("outputs", outputs.asJava)  // Dataset entity
     entity
   }
+
+  def etlProcessToEntity(inputs: List[AtlasEntity],
+                        outputs: List[AtlasEntity],
```
Comment (jerryshao) [inline spark-atlas-connector/src/main/scala/com/hortonworks/spark/atlas/types/internal.scala]: 4 spaces indent

File: spark-atlas-connector/src/main/scala/com/hortonworks/spark/atlas/types/internal.scala [hunk 8]
```diff
@@ -18,18 +18,20 @@
 package com.hortonworks.spark.atlas.types
 
 import scala.collection.JavaConverters._
-
```
Comment (jerryshao) [inline spark-atlas-connector/src/main/scala/com/hortonworks/spark/atlas/types/internal.scala]: revert this change.

File: spark-atlas-connector/src/main/scala/com/hortonworks/spark/atlas/types/internal.scala [hunk 9]
```diff
@@ -18,18 +18,20 @@
...
 import org.apache.atlas.AtlasClient
 import org.apache.atlas.model.instance.AtlasEntity
 import org.apache.spark.sql.catalyst.catalog.{CatalogDatabase, CatalogStorageFormat, CatalogTable}
 import org.apache.spark.sql.execution.QueryExecution
 import org.apache.spark.sql.types.StructType
 import org.apache.spark.ml.{Pipeline, PipelineModel}
-
```
Comment (jerryshao) [inline spark-atlas-connector/src/main/scala/com/hortonworks/spark/atlas/types/internal.scala]: ditto.

File: spark-atlas-connector/src/main/scala/com/hortonworks/spark/atlas/types/internal.scala [hunk 10]
```diff
@@ -18,18 +18,20 @@
...
 import com.hortonworks.spark.atlas.utils.{Logging, SparkUtils}
 
+import scala.collection.mutable
```
Comment (jerryshao) [inline spark-atlas-connector/src/main/scala/com/hortonworks/spark/atlas/types/internal.scala]: adjust the import ordering.

**Comment by jerryshao** [comment]
CC @dongjoon-hyun to also take a review.

**Comment by dongjoon-hyun** [comment]
Thank you for pinging me, @jerryshao .

File: spark-atlas-connector/src/main/scala/com/hortonworks/spark/atlas/SparkAtlasEventTracker.scala [hunk 11]
```diff
@@ -115,4 +115,4 @@ class SparkAtlasEventTracker(atlasClient: AtlasClient, atlasClientConf: AtlasCli
         false
     }
   }
-}
\ No newline at end of file
+}
```
Comment (dongjoon-hyun) [inline spark-atlas-connector/src/main/scala/com/hortonworks/spark/atlas/SparkAtlasEventTracker.scala]: Let's revert this change on this file, too.

File: spark-atlas-connector/src/main/scala/com/hortonworks/spark/atlas/ml/MLPipelineEventProcessor.scala [hunk 12]
```diff
@@ -154,53 +159,16 @@ class MLPipelineEventProcessor(
...
-          atlasClient.createEntities(Seq(modelDirEntity, modelEntity, transformEntity)
-            ++ tableEntities2.head ++ tableEntities3)
-
-          cachedObjects.remove(s"${uid}_modelEntity")
-          cachedObjects.remove(s"${uid}_modelDirEntity")
-          logInfo(s"Created transFormEntity ${transformEntity.getGuid}")
-        } else {
-          logInfo(s"Transform Entity is already created")
-        }
+        logInfo(s"Cache for TransformEvent " + uid)
```
Comment (dongjoon-hyun) [inline spark-atlas-connector/src/main/scala/com/hortonworks/spark/atlas/ml/MLPipelineEventProcessor.scala]: `logInfo("Cache for TransformEvent " + uid)` or `logInfo(s"Cache for TransformEvent $uid")`?

File: spark-atlas-connector/src/main/scala/com/hortonworks/spark/atlas/sql/CommandsHarvester.scala [hunk 13]
```diff
@@ -95,22 +111,40 @@ object CommandsHarvester extends AtlasEntityUtils with Logging {
             case r: FileRelation => r.inputFiles.map(external.pathToEntity).toSeq
             case _ => Seq.empty
           }
+        case local: LocalRelation =>
+          logInfo("Local Relation")
```
Comment (dongjoon-hyun) [inline spark-atlas-connector/src/main/scala/com/hortonworks/spark/atlas/sql/CommandsHarvester.scala]: Do we have a test coverage for this code path? And, could you put a more meaningful message or remove this?

File: spark-atlas-connector/src/main/scala/com/hortonworks/spark/atlas/sql/CommandsHarvester.scala [hunk 14]
```diff
@@ -95,22 +111,40 @@ object CommandsHarvester extends AtlasEntityUtils with Logging {
...
+        "remoteUser" -> SparkUtils.currSessionUser(qd.qe),
+        "executionTime" -> qd.executionTime.toString,
+        "details" -> qd.qe.toString(),
+        "sparkPlanDescription" -> qd.qe.sparkPlan.toString())
+
+      // ml related cached object
+      if (internal.cachedObjects.contains("model_uid")) {
+
+        internal.updateMLProcessToEntity(inputTablesEntities, outputEntities, logMap)
+
```
Comment (dongjoon-hyun) [inline spark-atlas-connector/src/main/scala/com/hortonworks/spark/atlas/sql/CommandsHarvester.scala]: Sorry for nit-picking, but let's not put blank space like this line 138 and line 140. There are more occurences like this. :)

File: spark-atlas-connector/src/main/scala/com/hortonworks/spark/atlas/sql/CommandsHarvester.scala [hunk 15]
```diff
@@ -142,9 +176,28 @@ object CommandsHarvester extends AtlasEntityUtils with Logging {
...
```
Comment (dongjoon-hyun) [inline spark-atlas-connector/src/main/scala/com/hortonworks/spark/atlas/sql/CommandsHarvester.scala]: ditto.

File: spark-atlas-connector/src/main/scala/com/hortonworks/spark/atlas/ml/MLPipelineEventProcessor.scala [hunk 16]
```diff
@@ -130,14 +132,17 @@ class MLPipelineEventProcessor(
             List(pipelineEntity, tableEntities.head.head),
             List(modelEntity))
 
-          atlasClient.createEntities(Seq(pipelineDirEntity, modelDirEntity,
-            pipelineEntity, modelEntity, fitEntity) ++ tableEntities.head)
+          // TODO add more information
```
Comment (dongjoon-hyun) [inline spark-atlas-connector/src/main/scala/com/hortonworks/spark/atlas/ml/MLPipelineEventProcessor.scala]: Please create a concrete issue for this and add issue ID, `TODO(SAC-XXX)`.

File: spark-atlas-connector/src/main/scala/org/apache/spark/sql/kafka010/atlas/KafkaHarvester.scala [hunk 17]
```diff
@@ -50,11 +48,16 @@ object KafkaHarvester extends AtlasEntityUtils with Logging {
...
     var destTopic = None: Option[String]
     try {
-      destTopic = node.getClass.getMethod("topic").invoke(node).asInstanceOf[Option[String]]
+      val topicField = node.getClass.getDeclaredField("topic")
+      topicField.setAccessible(true)
+      destTopic = topicField.get(node).asInstanceOf[Option[String]]
     } catch {
       case e: NoSuchMethodException =>
         logDebug(s"Can not get topic, so can not create Kafka topic entities: ${qd.qe}")
+      case e: Exception =>
```
Comment (dongjoon-hyun) [inline spark-atlas-connector/src/main/scala/org/apache/spark/sql/kafka010/atlas/KafkaHarvester.scala]: `case e: _ =>` ?

File: spark-atlas-connector/src/main/scala/com/hortonworks/spark/atlas/sql/CommandsHarvester.scala (hunk 13 above)
Comment (merlintang) [reply spark-atlas-connector/src/main/scala/com/hortonworks/spark/atlas/sql/CommandsHarvester.scala]: This case is used when Spark Pipeline model is saved. Because we already get enough information for Spark ML related entity. We just logInfo this tracking would be enough.

Comment (dongjoon-hyun) [reply spark-atlas-connector/src/main/scala/com/hortonworks/spark/atlas/sql/CommandsHarvester.scala]: Then, let's remove this.

Comment (merlintang) [reply spark-atlas-connector/src/main/scala/com/hortonworks/spark/atlas/sql/CommandsHarvester.scala]: without this, the tracker would throw a unknown message, this would be an annoying message. I am fixing your other comment, thanks for your kind help.

File: spark-atlas-connector/src/main/scala/com/hortonworks/spark/atlas/SparkAtlasEventTracker.scala (hunk 11 above)
Comment (merlintang) [reply spark-atlas-connector/src/main/scala/com/hortonworks/spark/atlas/SparkAtlasEventTracker.scala]: error file=/Users/mtang/Dev/spark/spark-atlas/spark-atlas-connector/spark-atlas-connector/src/main/scala/com/hortonworks/spark/atlas/SparkAtlasEventTracker.scala message=File must end with newline character

File: spark-atlas-connector/src/main/scala/com/hortonworks/spark/atlas/ml/MLPipelineEventProcessor.scala (hunk 1 above)
Comment (merlintang) [reply spark-atlas-connector/src/main/scala/com/hortonworks/spark/atlas/ml/MLPipelineEventProcessor.scala]: Thanks Jerry's review. 
1. For the global cache, this would not be a big issue since we only track ml related pipeline information. one spark application would have more than 1k pipeline related action. 

2. For the sql part, if uses do not have the ml job, the sql would run independently.

**Comment by merlintang** [comment]
@jerryshao and @dongjoon-hyun  can you review the update code?

**Comment by merlintang** [comment]
I will add test cases.

**Comment by merlintang** [comment]
if no further comments, I would like to merge this PR. @jerryshao @dongjoon-hyun
